src/main/java/edu/ucla/library/prl/harvester/Institution.java

```diff
@@ -0,0 +1,253 @@
+public final class Institution {
+
+    /**
+     * The JSON key for the name.
+     */
+    static final String NAME = "name";
+
+    /**
+     * The JSON key for the description.
+     */
+    static final String DESCRIPTION = "description";
+
+    /**
+     * The JSON key for the location.
+     */
+    static final String LOCATION = "location";
+
+    /**
+     * The JSON key for the email.
+     */
+    static final String EMAIL = "email";
+
+    /**
+     * The JSON key for the phone.
+     */
+    static final String PHONE = "phone";
+
+    /**
+     * The JSON key for the web contact.
+     */
+    static final String WEB_CONTACT = "webContact";
+
+    /**
+     * The JSON key for the website.
+     */
+    static final String WEBSITE = "website";
+
+    /**
+     * The logger.
+     */
+    private static final Logger LOGGER = LoggerFactory.getLogger(Institution.class, MessageCodes.BUNDLE);
+
+    /**
+     * Parses and formats phone numbers.
+     */
+    private static final PhoneNumberUtil PHONE_NUMBER_UTIL = PhoneNumberUtil.getInstance();
+
+    /**
+     * The institution's name.
+     */
+    private final String myName;
+
+    /**
+     * The institution's description.
+     */
+    private final String myDescription;
+
+    /**
+     * The institution's human-readable location.
+     */
+    private final String myLocation;
+
+    /**
+     * The optional email address contact of the institution.
+     */
+    private final Optional<InternetAddress> myEmail;
+
+    /**
+     * The optional phone number contact of the institution.
+     */
+    private final Optional<PhoneNumber> myPhone;
+
+    /**
+     * The optional web contact of the institution.
+     */
+    private final Optional<URL> myWebContact;
+
+    /**
+     * The institiution's website.
+     */
+    private final URL myWebsite;
+
+    /**
+     * Instantiates an institution.
+     * <p>
+     * At least one of {@code anEmail}, {@code aPhone}, or {@code aWebContact} must be provided.
+     *
+     * @param aName The institution's name
+     * @param aDescription The institution's description
+     * @param aLocation The institution's human-readable location
+     * @param anEmail The optional email address contact of the institution
+     * @param aPhone The optional phone number contact of the institution
+     * @param aWebContact The optional web contact of the institution
+     * @param aWebsite The institiution's website
+     */
+    @SuppressWarnings("PMD.AvoidThrowingNullPointerException")
+    public Institution(final String aName, final String aDescription, final String aLocation,
+            final InternetAddress anEmail, final PhoneNumber aPhone, final URL aWebContact, final URL aWebsite) {
+        myName = Objects.requireNonNull(aName);
+        myDescription = Objects.requireNonNull(aDescription);
+        myLocation = Objects.requireNonNull(aLocation);
+
+        if (anEmail == null && aPhone == null && aWebContact == null) {
```

Comment:
Oh, because DataObject doesn't support Optional(s)?

Reply:
One thing we haven't explored yet, to my knowledge, is using a mapper for types codegen doesn't support: https://github.com/eclipse-vertx/vertx-codegen#json-mappers

Reply:
Well, _providing more than one_ is optional. Since the optional-ity of each one depends on whether the others are provided, the way it is seems clearer to me.

Reply:
That one is required seems like a separate thing to me. Using Optionals, one could just do:
```
if (anEmail.isEmpty() && aPhone.isEmpty() && aWebContact.isEmpty()) {
    // Exception
} else {
     
}
```
I do agree with the AvoidThrowingNullPointerException warning, too. This would, IMHO, be a good case for a specialized exception (e.g. InvalidInstitutionException or something b/c an institution isn't considered valid without one form of contact).

Reply:
Fwiw, Eclipse (for me) also shortens the if/else to:
```
        if (anEmail == null && aPhone == null && aWebContact == null) {
            throw new NullPointerException(LOGGER.getMessage(MessageCodes.PRL_002));
        }

        myEmail = Optional.ofNullable(anEmail);
        myPhone = Optional.ofNullable(aPhone);
        myWebContact = Optional.ofNullable(aWebContact);
```
since the else isn't needed.

Reply:
So for something like:

```
Institution(String name, String description, String location, Optional<InternetAddress> email, Optional<PhoneNumber> phone, Optional<URL> webContact, URL website)
```

doesn't that seem to suggest that each one is optional, independent of the others?

I think if

```java
Institution("n", "d", "l", Optional.empty(), Optional.empty(), Optional.empty(), website);
```

compiles but is "invalid", that would be more surprising than if

```java
Institution("n", "d", "l", null, null, null, website);
```

compiles but is invalid.

Although, IMO neither are ideal, since they don't clearly express the "at least one of" idea; [this type](http://www.functionaljava.org/javadoc/5.0/functionaljava/fj/data/Either3.html) expresses the concept of "exactly one of", but that's not exactly right.

I wonder if a new type `ContactMethods` would help.

Reply:
Re: "doesn't that seem to suggest that each one is optional, independent of the others?" I think it suggests each is optional (without making any requirements about one must be present), which is a step above just args that may be null (which have no indication that they're not, each, required).

Reply:
I agree neither conveys the requirement in the API. There may be a better way?

Reply:
ContactMethods would be another way to do it too, yes.

Reply:
ContactMethods defined